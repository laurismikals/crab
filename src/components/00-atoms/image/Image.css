.image {
  background-color: #ffffff;
  position: relative;
  overflow: hidden;
}
.image img.image__noscript {
  opacity: 1;
  z-index: 1;
}
.image__loading {
  background-color: #cccccc;
  position: absolute;
  z-index: 0;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  animation-direction: alternate;
  animation-duration: 1s;
  animation-fill-mode: forwards;
  animation-iteration-count: infinite;
  animation-name: loadingPulse;
  animation-timing-function: ease-in-out;
}
@keyframes loadingPulse {
  from {
    opacity: 0.3;
  }
  to {
    opacity: 0.7;
  }
}
.image--no-loading .image__loading {
  animation: unset;
}
.image img {
  position: relative;
  z-index: 1;
  opacity: 0;
  transition: opacity 0.5s;
}
.image--loaded img {
  opacity: 1;
}
.image--aspect {
  float: left;
  width: 100%;
  position: relative;
}
.image--aspect img {
  position: absolute;
  top: -1px;
  left: -1px;
  width: calc(100% + 2px);
  height: calc(100% + 2px);
}
.image--cover img {
  width: calc(100% + 2px);
  height: calc(100% + 2px);
  object-fit: cover;
}
.no-objectfit .image--cover img,
.no-object-fit .image--cover img {
  position: absolute;
  top: 50%;
  left: 50%;
  width: auto;
  height: auto;
  min-width: 100%;
  min-height: 100%;
  transform: translateX(-50%) translateY(-50%);
  object-fit: cover;
}
.no-objectfit.no-csstransforms .image--cover img {
  top: 0;
  left: 0;
}
.image__caption {
  position: absolute;
  z-index: 2;
  bottom: 0;
  left: 0;
  width: 100%;
  padding: 20px;
  text-align: right;
  color: #ffffff;
  background: linear-gradient(to bottom, rgba(68, 68, 68, 0) 0%, rgba(68, 68, 68, 1) 100%);
  display: none;
}
@media only screen and (min-width: 541px) {
  .image__caption {
    display: block;
  }
}
