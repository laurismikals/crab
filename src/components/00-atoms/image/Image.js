import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import LazyLoad from 'react-lazy-load';

import './Image.css';

export class Image extends PureComponent {
  static propTypes = {
    src: PropTypes.string.isRequired,
    alt: PropTypes.string.isRequired,
    x: PropTypes.number,
    y: PropTypes.number,
    className: PropTypes.string,
  };
  static defaultProps = {
    x: null,
    y: null,
    className: '',
  };
  state = {
    loaded: false,
  };
  componentWillReceiveProps(nextProps) {
    if (this.props.src !== nextProps.src) {
      this.setState({
        loaded: false,
      });
    }
  }
  loadHandler = () => {
    this.setState({
      loaded: true,
    });
  };
  render() {
    const {
      className,
      x,
      y,
      src,
      alt,
    } = this.props;

    const padding = x && y ? { padding: `0 0 ${(y * 100) / x}% 0` } : {};

    return (
      <figure
        className={`
          ${className}
          image
          ${x && y ? 'image--aspect' : ''}
          ${this.state.loaded ? 'image--loaded' : 'image--loading'}
        `}
        style={padding}
      >
        <LazyLoad
          offsetVertical={900}
        >
          <img
            src={src}
            alt={alt}
            onLoad={this.loadHandler}
          />
        </LazyLoad>
        <span className="image__loading" />
      </figure>
    );
  }
}

export default Image;
