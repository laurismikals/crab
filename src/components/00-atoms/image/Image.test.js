import React from 'react';
import Enzyme, { shallow } from 'enzyme';

import { Image } from './Image';

describe('Atom: Image', () => {
  const minProps = {
    src: 'test.jpg',
    alt: 'Alt text',
  };

  it('renders correctly with minimal props', () => {
    const tree = shallow(
      <Image {...minProps} />
    );
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with proportions provided', () => {
    const tree = shallow(
      <Image {...minProps} x={1} y={1} />
    );
    expect(tree).toMatchSnapshot();
  });

  it('adds class image--loaded when state is set to loaded: true', () => {
    const tree = mount(
      <Image {...minProps} />
    );

    tree.setState({loaded: true});

    expect(
      tree.find('.image').hasClass('image--loaded')
    ).toEqual(true);
  });

  it('changes state.loaded to false when src is changed', () => {
    const tree = mount(
      <Image {...minProps} />
    );

    tree.setProps({ src: 'test_2.jpg' });

    expect(
      tree.state().loaded
    ).toEqual(false);
  });

});
