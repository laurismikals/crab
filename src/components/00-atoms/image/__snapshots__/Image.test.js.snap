// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Atom: Image renders correctly with minimal props 1`] = `
<figure
  className="
          
          image
          
          image--loading
        "
  style={Object {}}
>
  <LazyLoad
    debounce={true}
    elementType="div"
    offset={0}
    offsetBottom={0}
    offsetHorizontal={0}
    offsetLeft={0}
    offsetRight={0}
    offsetTop={0}
    offsetVertical={900}
    throttle={250}
  >
    <img
      alt="Alt text"
      onLoad={[Function]}
      src="test.jpg"
    />
  </LazyLoad>
  <span
    className="image__loading"
  />
</figure>
`;

exports[`Atom: Image renders correctly with proportions provided 1`] = `
<figure
  className="
          
          image
          image--aspect
          image--loading
        "
  style={
    Object {
      "padding": "0 0 100% 0",
    }
  }
>
  <LazyLoad
    debounce={true}
    elementType="div"
    offset={0}
    offsetBottom={0}
    offsetHorizontal={0}
    offsetLeft={0}
    offsetRight={0}
    offsetTop={0}
    offsetVertical={900}
    throttle={250}
  >
    <img
      alt="Alt text"
      onLoad={[Function]}
      src="test.jpg"
    />
  </LazyLoad>
  <span
    className="image__loading"
  />
</figure>
`;
